spynnaker/pyNN/models/neuron/synaptic_manager.py

```diff
@@ -322,14 +323,20 @@ def get_sdram_usage_in_bytes(
             self, vertex_slice, machine_time_step, application_graph,
             app_vertex):
         in_edges = application_graph.get_edges_ending_at_vertex(app_vertex)
-        return (
-            self._get_synapse_params_size() +
-            self._get_synapse_dynamics_parameter_size(
-                vertex_slice, application_graph, app_vertex) +
-            self._get_synaptic_blocks_size(
-                vertex_slice, in_edges, machine_time_step) +
-            self.__poptable_type.get_master_population_table_size(in_edges) +
-            self._get_size_of_generator_information(in_edges))
+        # These values can be cross checked with the data_spec_text_files
+        # SYNAPSE_PARAMS = 2
+        region_2 = self._get_synapse_params_size()
+        # SYNAPSE_DYNAMICS = 5
+        region_5 = self._get_synapse_dynamics_parameter_size(
+            vertex_slice, application_graph, app_vertex)
+        # SYNAPTIC_MATRIX = 4
+        region_4 = self._get_synaptic_blocks_size(
+            vertex_slice, in_edges, machine_time_step)
+        # POPULATION_TABLE = 3
+        region_3 = self.__poptable_type.get_master_population_table_size(in_edges)
+        #  DIRECT_MATRIX = 10
+        region_10 = self._get_size_of_generator_information(in_edges)
+        return region_2 + region_5 + region_4 + region_3 + region_10
```

Comment:
It looks like the SDRAM calculation misses the size of the direct matrix; this looks like it could be a tricky calculation, and should technically be subtracted from the synaptic blocks size (or else this calculation of this should remove the direct synapses).  I think that the direct and normal synaptic matrices were once a single region, so the calculation was valid, but when they were split up, this error crept in.

To be complete, the regions required here are:
 - [x] SYNAPSE_PARAMS (2)
 - [x] POPULATION_TABLE (3)
 - [x] SYNAPTIC_MATRIX (4)
 - [x] SYNAPSE_DYNAMICS (5)
 - [x] CONNECTION_BUILDER (9)
 - [ ] DIRECT_MATRIX (10)

Reply:
It was exactly my intention in this rewrite to align the cost methods with the regions so that we could detect the differences.

If you say that it is easier or more accurate to estimate the size of regions SYNAPTIC_MATRIX and DIRECT_MATRIX in a single method that is fine.  It only really makes sense to do a subtraction if someone want to check the estimate less the direct matrix estimate against the actual.

Reply:
Unfortunately, the issue is that the space will be estimated only once (both the DIRECT_MATRIX and SYNAPTIC_MATRIX together) and then the memory will be reserved for the SYNAPTIC_MATRIX using this value, and then even more memory will be allocated for the DIRECT_MATRIX.  This is again hidden because it will then estimate use for the CONNECTION_BUILDER which it won't then use.

As an example, the following should show the error:
```
import spynnaker8 as p

p.setup(1.0)
pre = p.Population(1, p.SpikeSourcePoisson(rate=1))
post = p.Population(1, p.IF_curr_exp())
post.record("all")


def no_gen(self, weights, delays):
    return False


p.OneToOneConnector.generate_on_machine = no_gen

proj = p.Projection(pre, post, p.OneToOneConnector(), p.StaticSynapse())

p.run(100)
```

Note that this overrides the generate_on_machine of the OneToOneConnector to stop the extra memory being allocated there from hiding the issue here.

So what is needed is a way to get a size for the SYNAPTIC_MATRIX and DIRECT_MATRIX separately so that this can be used in both estimation and region creation.